xarray_schema/components.py

```diff
@@ -0,0 +1,204 @@
+
+    def __init__(self, dtype: np.typing.DTypeLike) -> None:
+        if dtype in [np.floating, np.integer, np.signedinteger, np.unsignedinteger, np.generic]:
+            self.dtype = dtype
+        else:
+            self.dtype = np.dtype(dtype)
+
+    def validate(self, dtype: np.typing.DTypeLike) -> None:
+        '''Validate dtype
+
+        Parameters
+        ----------
+        dtype : Any
+            Dtype of the DataArray.
+        '''
+        if not np.issubdtype(dtype, self.dtype):
+            raise SchemaError(f'dtype {dtype} != {self.dtype}')
+
+    @property
+    def json(self) -> str:
+        if isinstance(self.dtype, np.dtype):
+            return self.dtype.str
+        else:
+            # fallbacks
+            return getattr(self.dtype, '__name__', str(self.dtype))
+
+
+class DimsSchema(BaseSchema):
+
+    _json_schema = {'type': 'array', 'items': {'type': ['string', 'null']}}
+
+    def __init__(self, dims: DimsT) -> None:
+        self.dims = dims
+
+    def validate(self, dims: tuple) -> None:
+        '''Validate dimensions
+
+        Parameters
+        ----------
+        dims : Tuple[Union[str, None]]
+            Dimensions of the DataArray. `None` may be used as a wildcard value.
+        '''
+        if len(self.dims) != len(dims):
+            raise SchemaError(f'length of dims does not match: {len(dims)} != {len(self.dims)}')
+
+        for i, (actual, expected) in enumerate(zip(dims, self.dims)):
+            if expected is not None and actual != expected:
+                raise SchemaError(f'dim mismatch in axis {i}: {actual} != {expected}')
+
+    @property
+    def json(self) -> list:
+        return list(self.dims)
+
+
+class ShapeSchema(BaseSchema):
+
+    _json_schema = {'type': 'array', 'items': {'type': ['int', 'null']}}
+
+    def __init__(self, shape: ShapeT) -> None:
+        self.shape = shape
+
+    def validate(self, shape: tuple) -> None:
+        '''Validate shape
+
+        Parameters
+        ----------
+        shape : ShapeT
+            Shape of the DataArray. `None` may be used as a wildcard value.
+        '''
+        if len(self.shape) != len(shape):
+            raise SchemaError(
+                f'number of dimensions in shape ({len(shape)}) != da.ndim ({len(self.shape)})'
+            )
+
+        for i, (actual, expected) in enumerate(zip(shape, self.shape)):
+            if expected is not None and actual != expected:
+                raise SchemaError(f'shape mismatch in axis {i}: {actual} != {expected}')
+
+    @property
+    def json(self) -> list:
+        return list(self.shape)
+
+
+class NameSchema(BaseSchema):
+
+    _json_schema = {'type': 'string'}
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+
+    def validate(self, name: Hashable) -> None:
+        '''Validate name
+
+        Parameters
+        ----------
+        name : str, optional
+            Name of the DataArray. Currently requires an exact string match.
```

Comment:
Maybe temp until regex matching, but would comparing the .lower or .upper name help?